tests/unit/GenericGraph.cpp

```diff
@@ -617,3 +617,234 @@ BOOST_AUTO_TEST_CASE(MutableEdgeNodeRemovalTest) {
+  auto CUndirectedEdges = C->undirected_child_edges();
+  auto CUndirectedDistance = distance(CUndirectedEdges.begin(),
+                                      CUndirectedEdges.end());
+  Count = C->successorCount() + C->predecessorCount();
+  revng_check(1 == CSuccessorDistance);
+  revng_check(1 == CPredecessorDistance);
+  revng_check(2 == CUndirectedDistance);
+  revng_check(next(CUndirectedEdges.begin(), Count) == CUndirectedEdges.end());
+
+  auto CSuccessorIt = CSuccessorEdges.begin();
+  auto CEdgeView = *CSuccessorIt;
+  revng_check(CEdgeView.Neighbor == D);
+  revng_check(*CEdgeView.Label == 4);
+  CSuccessorIt++;
+  revng_check(CSuccessorIt == CSuccessorEdges.end());
+
+  auto CPredecessorIt = CPredecessorEdges.begin();
+  CEdgeView = *CPredecessorIt;
+  revng_check(CEdgeView.Neighbor == A);
+  revng_check(*CEdgeView.Label == 2);
+  CPredecessorIt++;
+  revng_check(CPredecessorIt == CPredecessorEdges.end());
+
+  auto CUndirectedIt = CUndirectedEdges.begin();
+  CEdgeView = *CUndirectedIt;
+  revng_check(CEdgeView.Neighbor == D);
+  revng_check(*CEdgeView.Label == 4);
+  CUndirectedIt++;
+  CEdgeView = *CUndirectedIt;
+  revng_check(CEdgeView.Neighbor == A);
+  revng_check(*CEdgeView.Label == 2);
+  CUndirectedIt++;
+  revng_check(CUndirectedIt == CUndirectedEdges.end());
+
+  auto DSuccessorEdges = D->successor_edges();
+  auto DSuccessorDistance = distance(DSuccessorEdges.begin(),
+                                     DSuccessorEdges.end());
+  auto DPredecessorEdges = D->predecessor_edges();
+  auto DPredecessorDistance = distance(DPredecessorEdges.begin(),
+                                       DPredecessorEdges.end());
+  auto DUndirectedEdges = D->undirected_child_edges();
+  auto DUndirectedDistance = distance(DUndirectedEdges.begin(),
+                                      DUndirectedEdges.end());
+  Count = D->successorCount() + D->predecessorCount();
+  revng_check(0 == DSuccessorDistance);
+  revng_check(2 == DPredecessorDistance);
+  revng_check(2 == DUndirectedDistance);
+  revng_check(next(DUndirectedEdges.begin(), Count) == DUndirectedEdges.end());
+
+  auto DSuccessorIt = DSuccessorEdges.begin();
+  revng_check(DSuccessorIt == DSuccessorEdges.end());
+
+  auto DPredecessorIt = DPredecessorEdges.begin();
+  auto DEdgeView = *DPredecessorIt;
+  revng_check(DEdgeView.Neighbor == B);
+  revng_check(*DEdgeView.Label == 3);
+  DPredecessorIt++;
+  DEdgeView = *DPredecessorIt;
+  revng_check(DEdgeView.Neighbor == C);
+  revng_check(*DEdgeView.Label == 4);
+  DPredecessorIt++;
+  revng_check(DPredecessorIt == DPredecessorEdges.end());
+
+  auto DUndirectedIt = DUndirectedEdges.begin();
+  DEdgeView = *DUndirectedIt;
+  revng_check(DEdgeView.Neighbor == B);
+  revng_check(*DEdgeView.Label == 3);
+  DUndirectedIt++;
+  DEdgeView = *DUndirectedIt;
+  revng_check(DEdgeView.Neighbor == C);
+  revng_check(*DEdgeView.Label == 4);
+  DUndirectedIt++;
+  revng_check(DUndirectedIt == DUndirectedEdges.end());
+}
```

Comment:
I think we're missing tests for weirder graphs.
- multiple edges between a given pair of node A->B
- loops and self-loops in the underlying directed graph

Reply:
And I would like to test the visits for them too.

Reply:
I added a test with a graph which has multiple edges between the same node pair https://github.com/revng/revng/blob/ff2b7e90cbf0ff106a7853aa468bfff9920cc855/tests/unit/GenericGraph.cpp#L728

and a graph with a main loop and a smaller self loop https://github.com/revng/revng/blob/ff2b7e90cbf0ff106a7853aa468bfff9920cc855/tests/unit/GenericGraph.cpp#L749

Reply:
:+1: